### PR TITLE
Add entry IDs to the memory queue

### PR DESCRIPTION
**Pull request by faec**
## What does this PR do?

Adds an incrementing ID to events in the memory queue, as described in https://github.com/elastic/elastic-agent-shipper/issues/27. This is needed by the shipper, to report to inputs whether their events have been sent yet.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

Related: https://github.com/elastic/elastic-agent-shipper/issues/27

### CODE REVIEW

File: libbeat/publisher/queue/memqueue/ringbuf.go [hunk 1]
```diff
@@ -52,10 +52,10 @@ type region struct {
...
 }
 
-type clientState struct {
-	seq   uint32        // event sequence number
-	state *produceState // the producer it's state used to compute and signal the ACK count
-}
+/*type producerEventState struct {
+	eventIndex uint64 // Order of the event in its originating producer, independent of queue ID
+	producer   *ackProducer
+}*/
```
Comment (rdner) [inline libbeat/publisher/queue/memqueue/ringbuf.go:58]: Is it supposed to be commented?

File: libbeat/publisher/queue/diskqueue/consumer.go [hunk 2]
```diff
@@ -86,10 +86,14 @@ func (batch *diskQueueBatch) Count() int {
 	return len(batch.frames)
 }
 
-func (batch *diskQueueBatch) Event(i int) interface{} {
+func (batch *diskQueueBatch) Entry(i int) interface{} {
 	return batch.frames[i].event
 }
 
+func (batch *diskQueueBatch) ID(i int) queue.EntryID {
+	return 0
```
Comment (rdner) [inline libbeat/publisher/queue/diskqueue/consumer.go:94]: So the disk queue would not support the incremental IDs? Or it's going to be implemented in a separate PR?

File: libbeat/publisher/queue/memqueue/broker.go [hunk 3]
```diff
@@ -380,10 +388,14 @@ func (b *batch) Count() int {
...
 
-func (b *batch) Event(i int) interface{} {
+func (b *batch) Entry(i int) interface{} {
 	return b.entries[i].event
 }
 
+func (b *batch) ID(i int) queue.EntryID {
+	return b.entries[i].id
+}
+
```
Comment (rdner) [inline libbeat/publisher/queue/memqueue/broker.go:398]: Don't we need a boundary check and a returned error in exported functions like this? It should not panic if the `i` is out of range, I think?

File: libbeat/publisher/queue/memqueue/eventloop.go [hunk 4]
```diff
@@ -143,16 +145,18 @@ func (l *directEventLoop) handleMetricsRequest(req *metricsRequest) {
...
-		l.buf.insert(req.event, clientState{})
-	} else if st.cancelled {
+	if req.producer != nil && req.producer.state.cancelled {
 		reportCancelledState(log, req)
 	} else {
-		l.buf.insert(req.event, clientState{
-			seq:   req.seq,
-			state: st,
-		})
+		if req.resp != nil {
```
Comment (rdner) [inline libbeat/publisher/queue/memqueue/eventloop.go:151]: I think the comment to this `insert` function is outdated since it does not actually return anything and I think it needs a comment inside the function to explain why `req.resp` can be `nil` and how we handle it.

File: libbeat/publisher/queue/memqueue/eventloop.go [hunk 5]
```diff
@@ -203,49 +207,25 @@ func (l *directEventLoop) processACK(lst chanList, N int) {
...
-		}
-
-		count := (client.seq - client.state.lastACK)
-		if count == 0 || count > math.MaxUint32/2 {
-			// seq number comparison did underflow. This happens only if st.seq has
-			// already been acknowledged
-			// log.Debug("seq number already acked: ", st.seq)
-
-			client.state = nil
+		idx := (firstIndex + i) % len(entries)
```
Comment (rdner) [inline libbeat/publisher/queue/memqueue/eventloop.go:214]: It's not obvious what's going on with indexing here, could you please add a comment explaining this?

Also, I don't see where `len(entries)` is changed in the loop, it might worth to just pre-calculate this into `i` when the loop starts.

File: libbeat/publisher/queue/memqueue/eventloop.go [hunk 6]
```diff
@@ -203,49 +207,25 @@ func (l *directEventLoop) processACK(lst chanList, N int) {
...
+		entry := &entries[idx]
+
+		producer := entry.producer
+		entry.producer = nil
```
Comment (rdner) [inline libbeat/publisher/queue/memqueue/eventloop.go:218]: Is this how we mark that we have visited this entry before?

File: libbeat/publisher/queue/memqueue/produce.go [hunk 7]
```diff
@@ -42,66 +41,83 @@ type openState struct {
...
-	return p.updSeq(p.openState.tryPublish(p.makeRequest(event)))
-}
-
-func (p *ackProducer) updSeq(ok bool) bool {
-	if ok {
-		p.seq++
+func (p *ackProducer) Publish(event interface{}) (queue.EntryID, bool) {
+	id, published := p.openState.publish(p.makePushRequest(event))
+	if published {
+		p.producedCount++
```
Comment (rdner) [inline libbeat/publisher/queue/memqueue/produce.go:110]: Is it thread safe to just increment the counter in both `Publish` and `TryPublish`?

File: libbeat/publisher/queue/diskqueue/consumer.go (hunk 2 above)
Comment (faec) [reply libbeat/publisher/queue/diskqueue/consumer.go:94]: Yes, this is just for the memory queue, disk queue will be another PR (the shipper doesn't support the disk queue yet although with Lee's pending PR that should change any day now)

File: libbeat/publisher/queue/memqueue/ringbuf.go (hunk 1 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/ringbuf.go]: Nope, it's supposed to be deleted :-)

File: libbeat/publisher/queue/memqueue/eventloop.go (hunk 4 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/eventloop.go:151]: `req.resp` actually can't be nil! That check is leftover from an earlier iteration where it could be, so I removed it.

File: libbeat/publisher/queue/memqueue/eventloop.go (hunk 5 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/eventloop.go:216]: `len(entries)` doesn't change, only `i` does. I've added more comments clarifying where this value is coming from, see if that makes more sense.

File: libbeat/publisher/queue/memqueue/eventloop.go (hunk 6 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/eventloop.go:224]: Yes (though it may already be nil if the producer doesn't accept ACK callbacks). I've added a comment explaining.

File: libbeat/publisher/queue/memqueue/produce.go (hunk 7 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/produce.go:110]: No, but neither is the code that this replaces. The queues enforce thread-safe access but a single producer is only supposed to be used by a single goroutine at a time (though now that you mention it I'm not sure the current shipper code respects that).

Comment (faec) [reply libbeat/publisher/queue/memqueue/produce.go:110]: (Thinking more: I think the shipper case is actually ok, because it doesn't use or need a producer callback, so it won't hit `ackProducer`, just `forgetfulProducer` which doesn't need to track anything.)

**Comment by faec** [comment]
Added some more bits around metrics reporting, since I realized this code reported event IDs to the producers and consumers but the shipper also needs a standalone call to get the current ID, which I've added to the metrics struct already provided by the memory queue.

**Comment by faec** [comment]
The existing tests pass, the new tests for the event ID feature still need work:
- they involve a `time.Sleep` which is undesirable. Not sure yet if I can remove it, though, since the memory queue propagates ACKs asynchronously.
- some of them don't pass yet, particularly with the buffered version. I'm debugging that now, not 100% sure yet whether the problem is in the implementation or tests.

File: libbeat/publisher/queue/memqueue/produce.go [hunk 8]
```diff
@@ -42,66 +41,83 @@ type openState struct {
 	events chan pushRequest
 }
 
+// producerID stores the order of events within a single producer, so multiple
+// event acknowledgement callbacks can be coalesced into a single call.
+// It is defined as an explicit type to reduce cross-confusion with the id
+// of an event in the queue itself, which is a queue.EntryID.
+// with the
```
Comment (cmacknz) [inline libbeat/publisher/queue/memqueue/produce.go:48]:
```go
// with the
```

The comment here isn't finished

File: libbeat/publisher/queue/memqueue/produce.go [hunk 9]
```diff
@@ -42,66 +41,83 @@ type openState struct {
...
 		p.state.dropCB = dropCB
 		return p
 	}
 	return &forgetfulProducer{broker: b, openState: openState}
 }
 
-func (p *forgetfulProducer) Publish(event interface{}) bool {
-	return p.openState.publish(pushRequest{event: event})
+func (p *forgetfulProducer) makePushRequest(event interface{}) pushRequest {
+	resp := make(chan queue.EntryID, 1)
```
Comment (cmacknz) [inline libbeat/publisher/queue/memqueue/produce.go:77]: We make a channel for each event? Is it worth making one for each batch? Could we make a version of `Publish` that accepts an entire event batch instead of an individual event?

File: libbeat/publisher/queue/memqueue/eventloop.go [hunk 10]
```diff
@@ -203,49 +216,34 @@ func (l *directEventLoop) processACK(lst chanList, N int) {
...
 
 	firstIndex := lst.front().start
-	// Position the index at the end of the block of ACKed events
-	idx := (firstIndex + N - 1) % len(entries)
 
-	total := 0
+	// We want to acknowledge N events starting at position firstIndex
+	// in the entries array.
+	// We iterate over the events from last to first, so we encounter the
+	// highest produer IDs first and can skip subsequent callbacks to the
```
Comment (cmacknz) [inline libbeat/publisher/queue/memqueue/eventloop.go:223]:
```suggestion
	// highest producer IDs first and can skip subsequent callbacks to the
```

File: libbeat/publisher/queue/memqueue/eventloop.go [hunk 11]
```diff
@@ -323,11 +321,28 @@ func (l *bufferingEventLoop) run() {
...
+		occupiedRead:     l.unackedEventCount,
+		oldestEntryID:    oldestEntryID,
+	}
 }
 
 func (l *bufferingEventLoop) handleInsert(req *pushRequest) {
-	if l.insert(req) {
+	if l.insert(req, l.nextEntryID) {
+		// Send back the new event id.
+		req.resp <- l.nextEntryID
```
Comment (cmacknz) [inline libbeat/publisher/queue/memqueue/eventloop.go:343]: With these response channels, is there any chance that the receiving side has been cancelled or is otherwise unresponsive? Do we need to handle cases where this send blocks unexpectedly?

File: libbeat/publisher/queue/memqueue/queue_test.go [hunk 12]
```diff
@@ -175,3 +175,98 @@ func TestAdjustInputQueueSize(t *testing.T) {
...
+			assert.Equal(t, id, queue.EntryID(i), "Entry ID should match publication order")
+		}
+
+		for i := 0; i < entryCount; i++ {
+			batch, err := q.Get(1)
+			assert.NilError(t, err, "Queue read should succeed")
+			assert.Equal(t, batch.Count(), 1, "Returned batch should have 1 entry")
+			assert.Equal(t, batch.ID(0), queue.EntryID(i), "Consumed entry IDs should be ordered the same as when they were produced")
+
+			time.Sleep(100 * time.Millisecond)
```
Comment (cmacknz) [inline libbeat/publisher/queue/memqueue/queue_test.go:196]: What is this delay for?

File: libbeat/publisher/queue/memqueue/eventloop.go (hunk 11 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/eventloop.go:351]: No (though I could add more comments):
- these channels are only accessible internally to the queue
- any channel-based API call immediately listens on the response channel once the request is sent
- the response channels have buffer size 1 so the response can't block the main queue goroutine.
- once the queue accepts a request it will always send a response

File: libbeat/publisher/queue/memqueue/queue_test.go (hunk 12 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/queue_test.go]: troubleshooting, it isn't really needed and I'll remove it once I get the last test passing

File: libbeat/publisher/queue/memqueue/produce.go (hunk 9 above)
Comment (faec) [reply libbeat/publisher/queue/memqueue/produce.go:77]: It's not impossible but it would significantly complicate things: right now the memory queue accepts publish requests iff it has at least one free event slot, because then it can definitely handle the request without blocking. Accepting multiple events per push request would mean that's no longer true, so we'd need to assemble the response over time, and both event loops would need more complicated state handling to spread the response across multiple iterations.

However, ephemeral channels are quite cheap and this pattern is already used heavily in both queues. (We could, though, probably cache a single response channel for each producer, since the producers themselves are only supposed to be used from a single goroutine?)

Comment (cmacknz) [reply libbeat/publisher/queue/memqueue/produce.go:76]: Thanks, the pattern from the shipper side is to try to publish entire batches instead of individual events. I was more interested in seeing if we could improve the API for that use case, but it sounds like it isn't worth it unless there is some evidence of a performance issue.

Comment (faec) [reply libbeat/publisher/queue/memqueue/produce.go:76]: Ah, in that case yes I'd say it's a possible future optimization but probably not worth the complexity until we see where the real bottlenecks are. (My current belief is that the channeled API calls are negligible compared to the actual queueing process, so our current approach of calling single-event publish in a for loop doesn't lose much. It also gives us free semi-fair balancing between different inputs -- in the simplest implementation, keeping published batches together for the queue call would block all other inputs while a large publish request was being processed.)

File: libbeat/publisher/queue/memqueue/queue_test.go [hunk 13]
```diff
@@ -175,3 +176,165 @@ func TestAdjustInputQueueSize(t *testing.T) {
 		assert.Equal(t, int(float64(mainQueue)*maxInputQueueSizeRatio), AdjustInputQueueSize(mainQueue, mainQueue))
 	})
 }
+
+func TestEntryIDs(t *testing.T) {
+	entryCount := 100
+
+	testForward := func(q queue.Queue) {
+		waiter := &producerACKWaiter{}
+		producer := q.Producer(queue.ProducerConfig{ACK: waiter.ack})
```
Comment (cmacknz) [inline libbeat/publisher/queue/memqueue/queue_test.go:185]: Is it worth having a test case for concurrent producers?

**Comment by cmacknz** [comment]
LGTM, I'll wait for @rdner to confirm that his comments have been addressed before approving.

**Comment by cmacknz** [comment]
Would also be good to get @leehinman to look at this, I've pinged him on Slack.

File: libbeat/publisher/queue/diskqueue/consumer.go (hunk 2 above)
Comment (leehinman) [inline libbeat/publisher/queue/diskqueue/consumer.go:94]: Am I correct in assuming a "real" event ids for the disk queue would be in a follow on PR?  If so can we open an issue and make sure it is marked as "related" in this PR?

**Review by leehinman — APPROVED** [review]
In a follow on PR could we get a doc with some high level design docs/diagrams.

File: libbeat/publisher/queue/diskqueue/consumer.go (hunk 2 above)
Comment (faec) [reply libbeat/publisher/queue/diskqueue/consumer.go:94]: Yes, the issue is https://github.com/elastic/elastic-agent-shipper/issues/27
